atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    @VisibleForTesting
+    SequenceAndBound getAgreedState(long seq) {
+        final Optional<SequenceAndBound> state = getLearnedState(seq);
+        if (state.isPresent()) {
+            return state.get();
+        }
+
+        // In the common case seq - 1 will be agreed upon before seq is prepared.
+        Optional<SequenceAndBound> lastState = getLearnedState(seq - 1);
+        if (!lastState.isPresent()) {
+            // We know that even in the case of a truncate, seq - 2 will always be agreed upon.
+            SequenceAndBound forced = forceAgreedState(seq - 2, null);
+            lastState = Optional.of(forceAgreedState(seq - 1, forced.getBound()));
+        }
+
+        return forceAgreedState(seq, lastState.get().getBound());
+    }
+
+    @VisibleForTesting
+    SequenceAndBound forceAgreedState(long seq, @Nullable Long oldState) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L);
+        }
+
+        Optional<SequenceAndBound> state = getLearnedState(seq);
+        if (state.isPresent()) {
+            return state.get();
+        }
+
+        while (true) {
+            try {
+                byte[] value = proposer.propose(seq, oldState == null ? null : PtBytes.toBytes(oldState));
+                // propose must never be null.  We only pass in null for things we know are agreed upon already.
+                Preconditions.checkNotNull(value, "Proposed value can't be null, but was in sequence %s", seq);
+                return ImmutableSequenceAndBound.of(seq, PtBytes.toLong(value));
+            } catch (PaxosRoundFailureException e) {
+                log.info("failed during propose", e);
+                long backoffTime = getRandomBackoffTime();
+                try {
+                    Thread.sleep(backoffTime);
+                } catch (InterruptedException e1) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private Optional<SequenceAndBound> getLearnedState(long seq) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return Optional.of(ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L));
+        }
+        List<PaxosLong> responses = PaxosQuorumChecker.collectQuorumResponses(
+                ImmutableList.copyOf(learners),
+                l -> getLearnedValue(seq, l),
+                1,
+                executor,
+                PaxosQuorumChecker.DEFAULT_REMOTE_REQUESTS_TIMEOUT_IN_SECONDS,
+                true);
+        if (responses.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(ImmutableSequenceAndBound.of(seq, responses.iterator().next().getValue()));
+    }
+
+    private static PaxosLong getLearnedValue(long seq, PaxosLearner learner) {
+        PaxosValue value = learner.getLearnedValue(seq);
+        if (value == null) {
+            throw new NoSuchElementException();
```

Comment:
Okay, added `Tried to get a learned value for sequence number '%d' which didn't exist`. Unfortunately there doesn't seem to be an easy way to, from just the PaxosLearner itself, identify where this happened.